Elex 2766 separate out conformalization models

lennybronner opened the pull request:
## Description
In order to prepare for the Bootstrap model (which does not use conformalization), I had to change the inheritance structure of our models so far. We now have a `ConformalElectionModel` which inherits from the `BaseElectionModel` and the `GaussianElectionModel` and `NonparametricElectionModel` inherit from that. This means we can re-use the aggregation code from the `BaseElectionModel` in the upcoming bootstrap model without running into issues with a non-existent conformalization set. 

We now only deal with the conformalization data in the client if we are using a `ConformalElectionModel` rather than always, which is what we were doing previously.

I also added type hints and updated the unit tests correspondingly.

## Jira Ticket
[elex-2766 ](https://arcpublishing.atlassian.net/browse/ELEX-2766)

## Test Steps
Unit tests are updated, so `tox` works.

Also both old models work like normal:
```
elexmodel 2020-11-03_USA_G --office_id=P --estimands=dem --geographic_unit_type=county --pi_method=gaussian --percent_reporting=20 --model_parameters="{'lambda_': 1, 'beta': 4, 'winsorize': True}" --aggregates=postal_code --aggregates=county_classification --aggregates=unit
```
and 
```
elexmodel 2020-11-03_USA_G --office_id=P --estimands=dem --geographic_unit_type=county --pi_method=nonparametric --percent_reporting=20 --model_parameters="{'lambda_': 1, 'robust': True}"
```

dmnapolitano commented on src/elexmodel/models/BaseElectionModel.py:
```diff
@@ -1,94 +1,40 @@
-
-        # round since we don't need the artificial precision
-        return preds.round(decimals=0)
+        pass
```
Is the idea that the `BaseElectionModel` can't be instantiated and one has to use a subclass instead?  Because in that case, I'd recommend creating it as an abstract base class and have this (and similar) method(s) return `NotImplemented` so that an exception is raised: https://docs.python.org/3/library/abc.html 🤔

lennybronner replied:
yes, that is the idea! That's a great idea, will do -- thanks!

dmnapolitano commented on src/elexmodel/models/ConformalElectionModel.py:
```diff
@@ -0,0 +1,194 @@
+        """
+        Returns conformalization data at the aggregate level
+        """
+        pass
```
Why are these two methods just pass-throughs? 🤔

lennybronner replied:
same as above. will change to NotImplemented

lennybronner commented on src/elexmodel/models/BaseElectionModel.py:
```diff
@@ -1,94 +1,40 @@
-        X = df_X.values
-        y = df_y.values
-        weights = weights.values
+        return 10
```
In effect the number has to be greater than the number of features + fixed effects (with some slight modification in the case of conformalization). So in that case 10 is entirely arbitrary. I will change it to NotImplemented like your suggestion below!

lennybronner replied:
I think that would be ideal, but we don't have access to the final design matrix in this function so we don't actually know the number of columns 🙃

dmnapolitano replied:
Oh lol 😭

lennybronner commented:
> A couple of questions and suggestions but otherwise, I think this looks great! 🎉

Ok, I updated the two new models into abstract base classes and added/updated/moved unit tests correspondingly!
